schemachanger: use bazel test sharding for comparator tests

rafiss opened the pull request:
The tests previously worked by creating a logic test corpus then running
all those statements. However, that approach made it hard to parallelize
the tests using bazel test sharding.

Now, a test function is generated for each logic test file. Bazel will
take care of sharding all these tests.

fixes https://github.com/cockroachdb/cockroach/issues/111978
Release note: None

rafiss committed:
schemachanger: use bazel test sharding for comparator tests

The tests previously worked by creating a logic test corpus then running
all those statements. However, that approach made it hard to parallelize
the tests using bazel test sharding.

Now, a test function is generated for each logic test file. Bazel will
take care of sharding all these tests.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/113101)
<!-- Reviewable:end -->

rafiss committed:
cmd: remove generate-logictest-corpus

This is no longer needed since we don't generate corpus files anymore.

Release note: None

Xiang-Gu reviewed (commented):
I didn't understand the purpose of first commit where we collect more than one corpus files -- it seems like the third commit just deleted all the corpus collection logic. Do we still need commit 1?

Also, in commit 2 (which seems like the meat of this PR), you use template to generate one test for each file, how is this going to be helped in terms of speeding up/parallelization? 

Finally, the comparator testing is not backported to 23.1 (it only exists on master and 23.2 branch), so I don't think we should backport this to 23.1?

rafiss commented:
> I didn't understand the purpose of first commit where we collect more than one corpus files -- it seems like the third commit just deleted all the corpus collection logic. Do we still need commit 1?

You're right, we don't need it. That was my first attempt at speeding this up before I realized how we could use bazel's sharding.

> Also, in commit 2 (which seems like the meat of this PR), you use template to generate one test for each file, how is this going to be helped in terms of speeding up/parallelization?

Bazel will distribute each Test function in any package to a different shard. See https://cockroachlabs.slack.com/archives/C019A8NANH0/p1698258494931269 

> Finally, the comparator testing is not backported to 23.1 (it only exists on master and 23.2 branch), so I don't think we should backport this to 23.1?

I see, removing that label.

rickystewart commented on pkg/sql/schemachanger/comparator_generated_test.go:
```diff
@@ -0,0 +1,14316 @@
+	} else {
+		path = "../../../" + logicTestFile
+	}
+
```
So, first of all: you do not have to check this file in. Feel free to let it be `gitignore`'d. If you are going to check this in, I would encourage you to refactor out the massive duplication just so this file doesn't have have to be as colossal as it is.

rafiss replied:
i have removed the massive duplication

rickystewart replied:
Oh, awesome, thank you. Now it just slows down my browser instead of crashing it. 😛

Xiang-Gu commented:
Thanks Rafi for your replies. It looks good to me now pending Ricky's comment.

rafiss commented:
tftr!

bors r+
